File: zee/src/funcs.c
/* Miscellaneous Emacs functions reimplementation
   Copyright (c) 1997-2004 Sandro Sigala.
   Copyright (c) 2003-2006 Reuben Thomas.
   Copyright (c) 2004 David A. Capello.
   All rights reserved.

   This file is part of Zee.

   Zee is free software; you can redistribute it and/or modify it under
   the terms of the GNU General Public License as published by the Free
   Software Foundation; either version 2, or (at your option) any later
   version.

   Zee is distributed in the hope that it will be useful, but WITHOUT ANY
   WARRANTY; without even the implied warranty of MERCHANTABILITY or
   FITNESS FOR A PARTICULAR PURPOSE.  See the GNU General Public License
   for more details.

   You should have received a copy of the GNU General Public License
   along with Zee; see the file COPYING.  If not, write to the Free
   Software Foundation, Fifth Floor, 51 Franklin Street, Boston, MA
   02111-1301, USA.  */

#include "config.h"

#include <ctype.h>
#include <signal.h>
#include <unistd.h>

#include "main.h"
#include "extern.h"


DEF(suspend,
"\
Stop and return to superior process.\
")
{
  raise(SIGTSTP);
}
END_DEF

DEF(cancel,
"\
Cancel current command.\
")
{
  weigh_mark();
  minibuf_error(astr_new("Quit"));
  ok = FALSE;
}
END_DEF

DEF(edit_toggle_read_only,
"\
Change whether this buffer is visiting its file read-only.\
")
{
  buf->flags ^= BFLAG_READONLY;
}
END_DEF

DEF(auto_fill_mode,
"\
Toggle Auto Fill mode.\n\
In Auto Fill mode, inserting a space at a column beyond `fill_column'\n\
automatically breaks the line at a previous space.\
")
{
  buf->flags ^= BFLAG_AUTOFILL;
}
END_DEF

DEF_ARG(set_fill_column,
"\
Set the fill column.\n\
If an argument value is passed, set `fill_column' to that value,\n\
otherwise with the current column value.\
",
UINT(col, "Enter new fill column: "))
{
  set_variable(astr_new("fill_column"),
               astr_afmt("%lu", list_empty(l) ? buf->pt.o + 1 : col));
}
END_DEF

DEF(set_mark,
"\
Set mark where point is.\
")
{
  set_mark_to_point();
  minibuf_write(astr_new("Mark set"));
  anchor_mark();
}
END_DEF

DEF(exchange_point_and_mark,
"\
Put the mark where point is now, and point where the mark is now.\
")
{
  assert(buf->mark);
  swap_point(&buf->pt, &buf->mark->pt);
  anchor_mark();
  thisflag |= FLAG_NEED_RESYNC;
}
END_DEF

DEF(mark_whole_buffer,
"\
Put point at beginning and mark at end of buffer.\
")
{
  CMDCALL(end_of_buffer);
  CMDCALL(set_mark);
  CMDCALL(beginning_of_buffer);
}
END_DEF

static int quoted_insert_octal(int c1)
{
  int c2, c3;
  minibuf_write(astr_afmt("Insert octal character %d-", c1 - '0'));
  c2 = getkey();

  if (!isdigit(c2) || c2 - '0' >= 8) {
    insert_char(c1 - '0');
    insert_char(c2);
    return TRUE;
  }

  minibuf_write(astr_afmt("Insert octal character %d %d-", c1 - '0', c2 - '0'));
  c3 = getkey();

  if (!isdigit(c3) || c3 - '0' >= 8) {
    insert_char((c1 - '0') * 8 + (c2 - '0'));
    insert_char(c3);
    return TRUE;
  }

  insert_char((c1 - '8') * 64 + (c2 - '0') * 8 + (c3 - '0'));

  return TRUE;
}

DEF(quoted_insert,
"\
Read next input character and insert it.\n\
This is useful for inserting control characters.\n\
You may also type up to 3 octal digits, to insert a character with that code.\
")
{
  int c;

  minibuf_write(astr_new("Insert literal character: "));
  c = xgetkey(GETKEY_UNFILTERED, 0);

  if (isdigit(c) && c - '0' < 8)
    quoted_insert_octal(c);
  else
    insert_char(c);

  minibuf_clear();
}
END_DEF

DEF_ARG(repeat,
"\
Repeat a command a given number of times.\n\
FIXME: Make it work as advertised: get the command,\n\
and perform it as a single undo action.\
",
UINT(reps, "Enter repeat count: "))
{
  minibuf_clear();
}
END_DEF

DEF(back_to_indentation,
"\
Move point to the first non-whitespace character on this line.\
")
{
  buf->pt.o = 0;
  while (!eolp()) {
    if (!isspace(following_char()))
      break;
    CMDCALL(edit_navigate_forward_char);
  }
}
END_DEF


/***********************************************************************
			  Move through words
***********************************************************************/

DEF(forward_word,
"\
Move point forward one word.\
")
{
  int gotword = FALSE;

  for (;;) {
    while (!eolp()) {
      int c = following_char();
      if (!isalnum(c)) {
        if (gotword)
          break;
      } else
        gotword = TRUE;
      buf->pt.o++;
    }
    if (gotword)
      break;
    buf->pt.o = astr_len(buf->pt.p->item);
    if (!CMDCALL(edit_navigate_down_line)) {
      ok = FALSE;
      break;
    }
    buf->pt.o = 0;
  }
}
END_DEF

DEF(backward_word,
"\
Move backward until encountering the beginning of a word.\
")
{
  int gotword = FALSE;

  for (;;) {
    if (bolp()) {
      if (!CMDCALL(edit_navigate_up_line)) {
        ok = FALSE;
        break;
      }
      buf->pt.o = astr_len(buf->pt.p->item);
    }
    while (!bolp()) {
      int c = preceding_char();
      if (!isalnum(c)) {
        if (gotword)
          break;
      } else
        gotword = TRUE;
      buf->pt.o--;
    }
    if (gotword)
      break;
  }
}
END_DEF

DEF(mark_word,
"\
Set mark to end of current word.\
")
{
  CMDCALL(set_mark);
  if ((ok = CMDCALL(forward_word)))
    CMDCALL(exchange_point_and_mark);
}
END_DEF

DEF(mark_word_backward,
"\
Set mark to start of current word.\
")
{
  CMDCALL(set_mark);
  if ((ok = CMDCALL(backward_word)))
    CMDCALL(exchange_point_and_mark);
}
END_DEF

DEF(backward_paragraph,
"\
Move backward to start of paragraph.\
")
{
  while (is_empty_line() && CMDCALL(edit_navigate_up_line))
    ;
  while (!is_empty_line() && CMDCALL(edit_navigate_up_line))
    ;

  CMDCALL(beginning_of_line);
}
END_DEF

DEF(forward_paragraph,
"\
Move forward to end of paragraph.\
")
{
  while (is_empty_line() && CMDCALL(edit_navigate_down_line))
    ;
  while (!is_empty_line() && CMDCALL(edit_navigate_down_line))
    ;

  if (is_empty_line())
    CMDCALL(beginning_of_line);
  else
    CMDCALL(end_of_line);
}
END_DEF

DEF(mark_paragraph,
"\
Put point at beginning of this paragraph, mark at end.\n\
The paragraph marked is the one that contains point or follows point.\
")
{
  CMDCALL(forward_paragraph);
  CMDCALL(set_mark);
  CMDCALL(backward_paragraph);
}
END_DEF

/* FIXME: fill_paragraph undo goes bananas. */
DEF(fill_paragraph,
"\
Fill paragraph at or after point.\
")
{
  int i, start, end;
  Marker *m = point_marker();

  undo_save(UNDO_START_SEQUENCE, buf->pt, 0, 0);

  CMDCALL(forward_paragraph);
  end = buf->pt.n;
  if (is_empty_line())
    end--;

  CMDCALL(backward_paragraph);
  start = buf->pt.n;
  if (is_empty_line()) {  /* Move to next line if between two paragraphs. */
    CMDCALL(edit_navigate_down_line);
    start++;
  }

  for (i = start; i < end; i++) {
    CMDCALL(end_of_line);
    CMDCALL(delete_char);
    CMDCALL(delete_horizontal_space);
    insert_char(' ');
  }

  CMDCALL(end_of_line);
  while (get_goalc() > (size_t)get_variable_number(astr_new("fill_column")) + 1)
    fill_break_line();

  thisflag &= ~FLAG_DONE_CPCN;

  buf->pt = m->pt;
  remove_marker(m);

  undo_save(UNDO_END_SEQUENCE, buf->pt, 0, 0);
}
END_DEF

#define UPPERCASE		1
#define LOWERCASE		2
#define CAPITALIZE		3

static int setcase_word(int rcase)
{
  size_t i, size;
  int firstchar;

  if (!isalnum(following_char())) {
    if (!CMDCALL(forward_word))
      return FALSE;
    if (!CMDCALL(backward_word))
      return FALSE;
  }

  i = buf->pt.o;
  while (i < astr_len(buf->pt.p->item)) {
    if (!isalnum(*astr_char(buf->pt.p->item, (ptrdiff_t)i)))
      break;
    ++i;
  }
  if ((size = i - buf->pt.o) > 0)
    undo_save(UNDO_REPLACE_BLOCK, buf->pt, size, size);

  for (firstchar = TRUE;
       buf->pt.o < i;
       buf->pt.o++, firstchar = FALSE) {
    char c = *astr_char(buf->pt.p->item, (ptrdiff_t)buf->pt.o);

    if (isalpha(c)) {
      if (rcase == UPPERCASE)
        c = toupper(c);
      else if (rcase == LOWERCASE)
        c = tolower(c);
      else if (rcase == CAPITALIZE)
        c = firstchar ? toupper(c) : tolower(c);

      *astr_char(buf->pt.p->item, (ptrdiff_t)buf->pt.o) = c;
    } else if (!isdigit(c))
      break;
  }

  buf->flags |= BFLAG_MODIFIED;

  return TRUE;
}

DEF(downcase_word,
"\
Convert following word to lower case, moving over.\
")
{
  ok = setcase_word(LOWERCASE);
}
END_DEF

DEF(upcase_word,
"\
Convert following word to upper case, moving over.\
")
{
  ok = setcase_word(UPPERCASE);
}
END_DEF

DEF(capitalize_word,
"\
Capitalize the following word, moving over.\
")
{
  ok = setcase_word(CAPITALIZE);
}
END_DEF

DEF(execute_command,
"\
Read command or macro name, then call it.\n\
FIXME: Make it work non-interactively.\
")
{
  astr name;
  Command cmd;
  Macro *mp;

  if ((name = minibuf_read_command_name(astr_new("Execute command: ")))) {
    if ((cmd = get_command(name)))
      ok = cmd(list_new());
    else if ((mp = get_macro(name)))
      call_macro(mp);
    else
      ok = FALSE;
  } else
    ok = FALSE;
}
END_DEF

DEF(shell_command,
"\
Reads a line of text using the minibuffer and creates an inferior shell\n\
to execute the line as a command; passes the contents of the region as\n\
input to the shell command.\n\
If the shell command produces any output, it is inserted into the\n\
current buffer, overwriting the current region.\n\
FIXME: Use better-shell.c\
")
{
  astr ms;

  if ((ms = minibuf_read(astr_new("Shell command: "), astr_new(""))) == NULL)
    ok = CMDCALL(cancel);
  else if (astr_len(ms) == 0 || warn_if_no_mark())
    ok = FALSE;
  else {
    char tempfile[] = P_tmpdir "/" PACKAGE_NAME "XXXXXX";
    int fd = mkstemp(tempfile);

    if (fd == -1) {
      minibuf_error(astr_new("Cannot open temporary file"));
      ok = FALSE;
    } else {
      FILE *pipe;
      Region r;
      astr cmd = astr_new(""), as;

      assert(calculate_the_region(&r));
      as = copy_text_block(r.start, r.size);
      write(fd, astr_cstr(as), r.size);

      close(fd);

      cmd = astr_afmt("%s 2>&1 <%s", astr_cstr(ms), tempfile);

      if ((pipe = popen(astr_cstr(cmd), "r")) == NULL) {
        minibuf_error(astr_new("Cannot open pipe to process"));
        ok = FALSE;
      } else {
        astr out = astr_new(""), s;

        while (astr_len(s = astr_fgets(pipe)) > 0) {
          astr_cat(out, s);
          astr_cat(out, astr_new("\n"));
        }
        pclose(pipe);
        remove(tempfile);

#ifdef CURSES
        /* We have no way of knowing whether a sub-process caught a
           SIGWINCH, so raise one. */
        raise(SIGWINCH);
#endif

        undo_save(UNDO_START_SEQUENCE, buf->pt, 0, 0);
        calculate_the_region(&r);
        if (buf->pt.p != r.start.p
            || r.start.o != buf->pt.o)
          CMDCALL(exchange_point_and_mark);
        delete_nstring(r.size, &s);
        ok = insert_nstring(out);
        undo_save(UNDO_END_SEQUENCE, buf->pt, 0, 0);
      }
    }
  }
}
END_DEF

File: src/funcs.lua
-- Miscellaneous Emacs functions
--
-- Copyright (c) 2010-2012 Free Software Foundation, Inc.
--
-- This file is part of Zee.
--
-- This program is free software; you can redistribute it and/or modify it
-- under the terms of the GNU General Public License as published by
-- the Free Software Foundation; either version 3, or (at your option)
-- any later version.
--
-- This program is distributed in the hope that it will be useful, but
-- WITHOUT ANY WARRANTY; without even the implied warranty of
-- MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
-- General Public License for more details.
--
-- You should have received a copy of the GNU General Public License
-- along with this program.  If not, see <http://www.gnu.org/licenses/>.

Defun ("keyboard-quit",
[[
Cancel current command.
]],
  function ()
    deactivate_mark ()
    return minibuf_error ("Quit")
  end
)

Defun ("file-suspend",
[[
Stop editor and return to superior process.
]],
  function ()
    posix.raise (posix.SIGTSTP)
  end
)

Defun ("preferences-toggle-read-only",
[[
Change whether this buffer is visiting its file read-only.
]],
  function ()
    cur_bp.readonly = not cur_bp.readonly
  end
)

Defun ("preferences-toggle-wrap-mode",
[[
Toggle Auto Fill mode.
In Auto Fill mode, inserting a space at a column beyond `fill-column'
automatically breaks the line at a previous space.
]],
  function ()
    cur_bp.autofill = not cur_bp.autofill
  end
)

Defun ("edit-select-other-end",
[[
Put the mark where point is now, and point where the mark is now.
]],
  function ()
    if not cur_bp.mark then
      return minibuf_error ("No mark set in this buffer")
    end

    local tmp = get_buffer_pt (cur_bp)
    goto_offset (cur_bp.mark.o)
    cur_bp.mark.o = tmp
    activate_mark ()
    thisflag.need_resync = true
  end
)

function select_on ()
  set_mark ()
  activate_mark ()
end

Defun ("edit-select-on",
[[
Set the mark where point is.
]],
  function ()
    select_on ()
    minibuf_write ("Mark set")
  end
)

Defun ("edit-insert-quoted",
[[
Read next input character and insert it.
This is useful for inserting control characters.
]],
  function ()
    minibuf_write ("C-q-")
    insert_char (string.char (bit32.band (getkey_unfiltered (GETKEY_DEFAULT), 0xff)))
    minibuf_clear ()
  end
)

Defun ("edit-wrap-paragraph",
[[
Fill paragraph at or after point.
]],
  function ()
    local m = point_marker ()

    undo_start_sequence ()

    execute_function ("move-next-paragraph")
    if is_empty_line () then
      previous_line ()
    end
    local m_end = point_marker ()

    execute_function ("move-previous-paragraph")
    if is_empty_line () then -- Move to next line if between two paragraphs.
      next_line ()
    end

    while buffer_end_of_line (cur_bp, get_buffer_pt (cur_bp)) < m_end.o do
      execute_function ("move-end-line")
      delete_char ()
      execute_function ("delete-horizontal-space")
      insert_char (' ')
    end
    unchain_marker (m_end)

    execute_function ("move-end-line")
    while get_goalc () > tonumber (get_variable ("fill-column")) + 1 and fill_break_line () do end

    goto_offset (m.o)
    unchain_marker (m)

    undo_end_sequence ()
  end
)

local function pipe_command (cmd, tempfile)
  local cmdline = string.format ("%s 2>&1%s", cmd, tempfile and " <" .. tempfile or "")
  local pipe = io.popen (cmdline, "r")
  if not pipe then
    return minibuf_error ("Cannot open pipe to process")
  end

  local out = pipe:read ("*a")
  pipe:close ()

  if #out == 0 then
    minibuf_write ("(Shell command succeeded with no output)")
  elseif not warn_if_readonly_buffer () then
    local del = 0
    if cur_bp.mark_active then
      local r = calculate_the_region ()
      goto_offset (r.start)
      del = get_region_size (r)
    end
    replace_astr (del, AStr (out))
  end

  return true
end

local function minibuf_read_shell_command ()
  local ms = minibuf_read ("Shell command: ", "")

  if not ms then
    execute_function ("keyboard-quit")
    return
  end
  if ms == "" then
    return
  end

  return ms
end

Defun ("edit-shell-command",
[[
Execute string command in inferior shell with region as input.
The output is inserted in the buffer, replacing the region if any.
Return the exit code of command.
]],
  function (cmd)
    local ok = true

    if not cmd then
      cmd = minibuf_read_shell_command ()
    end

    if cmd then
      local rp = calculate_the_region ()
      local h, tempfile
      if rp then
        tempfile = os.tmpname ()

        h = io.open (tempfile, "w")
        if not h then
          ok = minibuf_error ("Cannot open temporary file")
        else
          local fd = posix.fileno (h)
          activate_mark ()
          local s = get_buffer_region (cur_bp, calculate_the_region ())
          local written, err = alien.default.write (fd, s.buf.buffer:topointer (), #s)
          if not written then
            ok = minibuf_error ("Error writing to temporary file: " .. err)
          end
        end
      end

      if ok then
        ok = pipe_command (cmd, tempfile)
      end
      if h then
        h:close ()
      end
      if rp then
        os.remove (tempfile)
      end
    end
    return ok
  end
)

local function move_paragraph (forward, line_extremum)
  repeat until not is_empty_line () or not forward ()
  repeat until is_empty_line () or not forward ()

  if is_empty_line () then
    execute_function ("move-start-line")
  else
    execute_function (line_extremum)
  end
end

Defun ("move-previous-paragraph",
[[
Move backward to start of paragraph.
]],
  function ()
    return move_paragraph (previous_line, "move-start-line")
  end
)

Defun ("move-next-paragraph",
[[
Move forward to end of paragraph.
]],
  function ()
    return move_paragraph (next_line, "move-end-line")
  end
)


-- Move through words
local function move_word (dir)
  local gotword = false
  repeat
    while not (dir > 0 and eolp or bolp) () do
      if get_buffer_char (cur_bp, get_buffer_pt (cur_bp) - (dir < 0 and 1 or 0)):match ("%w") then
        gotword = true
      elseif gotword then
        break
      end
      move_char (dir)
    end
  until gotword or not move_char (dir)
  return gotword
end

Defun ("move-next-word",
[[
Move point forward one word.
]],
  function ()
    return move_word (1)
  end
)

Defun ("move-previous-word",
[[
Move backward until encountering the end of a word.
]],
  function ()
    return move_word (-1)
  end
)
